api

```diff
@@ -2919,16 +2895,19 @@ send_error_report() { #non-interactively send a Pi-Apps error log file to the Bo
   
 }
 
-diagnose_apps() { #Given a list of apps that failed to install/uninstall, loop through each error log, diagnose it, and provide a "Send report" button if applicable.
-  local failed_apps="$1"
+diagnose_apps() { #Given a list of action;app that failed to install/uninstall, loop through each error log, diagnose it, and provide a "Send report" button if applicable.
+  local failure_list="$1"
   local IFS=$'\n'
   
-  local num_lines="$(grep . <<<"$failed_apps" | wc -l)"
-  local i=1 #counter to track which failed_app in the list is current
-  local app
+  local num_lines="$(grep . <<<"$failure_list" | wc -l)"
+  local i=1 #counter to track which line in the list is current
+  local line
   local button
   
-  for app in $failed_apps ;do
+  for line in $failure_list ;do
+    local action="$(echo "$line" | awk -F';' '{print $1}')"
+    local app="$(echo "$line" | awk -F';' '{print $2}')"
+    
```

Comment:
I think that `diagnose_apps` has had a bug since it was introduced. I went through the history tracing back where it was located (it used to be in manage) and all its iterations and found when it was added as part of https://github.com/Botspot/pi-apps/pull/1580 and it seems like "action" was never set on a per app basis. If it was set at all, it was always inherited from the last time "action" was set for every app in the for loop (so it was likely to be incorrect). So we were almost always misdiagnosing apps.

Reply:
You are correct. I think the bug was a result of this function originally being run once per app, and later it was changed to run for multiple failed apps. Whatever the cause, it is fixed now.

I think the reason why nobody reported this issue is because most people will update a group of apps, or install several. $action is likely to be the same for all apps at a time.